docs/apps/rbac/for_clients.md

```diff
@@ -0,0 +1,114 @@
+## Using DAB RBAC as an API Client
+
+This section tells how to use the API endpoints as a client;
+what requests to make and how to build request data.
+
+You need a server running a Django project that uses this system.
+Use test_app in this repo for a demo, see `test_app/README.md` for bootstrap.
+http://127.0.0.1:8000/admin/ allows login as admin user (password is "password").
+
+### Create Custom Role (Definition)
+
+After logging in to test_app/, you can visit this endpoint to get the role definition API.
+
+http://127.0.0.1:8000/api/v1/role_definitions/
+
+Perform an OPTIONS request, and you will find this gives choices for `content_type`.
+Out of the choices a single `content_type` needs to be chosen for a role definition.
+
+Multiple permissions can be selected, and are accepted in the form of a list.
+To find out what permissions are valid for a role definition for a given `content_type`
+make a GET to `/api/v1/role_metadata/` and look up the type under "allowed_permissions".
+
+A POST to this endpoint will create a new role definition, example data:
+
+```json
+{
+    "permissions": ["view_inventory"],
+    "content_type": "aap.inventory",
+    "name": "View a single inventory",
+    "description": "custom role"
+}
+```
+
+Name can be any string that's not blank. Description can be any string.
+
+### Assigning a User a Role to an Object
+
+Select a role definition from `/api/v1/role_definitions/`, use the id as `role_definition`
+and a user from `/api/v1/users/` and use the id as `user`.
+Given the type of the role definition, check the available objects of that type,
+in this case `/api/v2/inventories/` and obtain the desired id, this will become `object_id`.
+
+With all 3 ids ready, POST to http://127.0.0.1:8000/api/v1/role_user_assignments/
+
+```json
+{
+    "role_definition": 3,
+    "object_id": 3,
+    "user": 3
+}
+```
```

Comment:
The json content template is:
{
    "object_id": "",
    "object_ansible_id": null,
    "role_definition": null,
    "user": null,
    "user_ansible_id": null
}
not sure if we should fix this

Reply:
AWX had a solution for this with a generic `update_raw_data` method

https://github.com/ansible/awx/blob/devel/awx/api/generics.py#L361-L372

But it's fairly messy to implement, because a flag has to be set on the serializer in the renderer class, which we don't even do anything custom to in DAB yet. So I'm not yet ready to dive into that customization.

I'm wondering if we could, instead, change the fields to allow null, and then do nothing when getting a null value.

Reply:
https://github.com/ansible/django-ansible-base/pull/468